File: zombicideandroid/src/main/java/cc/game/zombicide/android/ZClientMgr.java
package cc.game.zombicide.android;

import android.content.DialogInterface;
import android.util.Log;

import java.util.ArrayList;
import java.util.List;

import cc.lib.logger.Logger;
import cc.lib.logger.LoggerFactory;
import cc.lib.net.GameClient;
import cc.lib.net.GameCommand;
import cc.lib.zombicide.ZGame;
import cc.lib.zombicide.ZQuests;
import cc.lib.zombicide.ZUser;
import cc.lib.zombicide.p2p.ZGameMP;
import cc.lib.zombicide.p2p.ZUserMP;
import cc.lib.zombicide.ui.UIZombicide;

/**
 * Created by Chris Caron on 7/28/21.
 */
class ZClientMgr extends ZMPCommon implements GameClient.Listener, ZMPCommon.CL {

    static Logger log = LoggerFactory.getLogger(ZClientMgr.class);

    final ZUser user;
    final GameClient client;
    CharacterChooserDialog playerChooser = null;

    ZClientMgr(ZombicideActivity activity, UIZombicide game, GameClient client, ZUser user) {
        super(activity, game);
        this.client = client;
        this.user = user;
        client.addListener(this);
    }

    void shutdown() {
        client.removeListener(this);
    }

    @Override
    public void onCommand(GameCommand cmd) {
        parseSVRCommand(cmd);
    }

    @Override
    public void onLoadQuest(ZQuests quest) {
        game.loadQuest(quest);
        game.boardRenderer.redraw();
    }

    @Override
    public void onInit(int color, int maxCharacters, List<Assignee> playerAssignments) {
        user.setColor(color);
        game.clearCharacters();
        game.clearUsersCharacters();

        List<Assignee> assignees = new ArrayList<>();
        for (ZombicideActivity.CharLock c : activity.charLocks) {
            Assignee a = new Assignee(c);
            int idx = playerAssignments.indexOf(a);
            if (idx >= 0) {
                Assignee aa = playerAssignments.get(idx);
                a.copyFrom(aa);
                if (color == a.color)
                    a.isAssingedToMe = true;
            }
            assignees.add(a);
            if (a.checked) {
                game.addCharacter(a.name);
                if (a.isAssingedToMe) {
                    user.addCharacter(a.name);
                }
            }
        }
        activity.runOnUiThread(new Runnable() {
            public void run() {
                playerChooser = new CharacterChooserDialog(activity, assignees, maxCharacters) {
                    @Override
                    protected void onAssigneeChecked(Assignee assignee, boolean checked) {
                        Log.d(TAG, "onAssigneeChecked: " + assignee);
                        GameCommand cmd = activity.clientMgr.newAssignCharacter(assignee.name,checked);
                        new CLSendCommandSpinnerTask(activity, ZMPCommon.SVR_ASSIGN_PLAYER) {
                            @Override
                            protected void onSuccess() {
                                assignee.checked = checked;
                                postNotifyUpdateAssignee(assignee);
                            }
                        }.execute(cmd);
                    }

                    @Override
                    protected void onStart() {
                        game.setClient(client);
                        client.sendCommand(newStartPressed());
                        activity.initGameMenu();
                    }
                };
                game.boardRenderer.redraw();
            }
        });
    }

    @Override
    public void onAssignPlayer(Assignee assignee) {
        Log.d("ZClientMgr", "onAssignPlayer: " + assignee);
        if (user.getColorId() == assignee.color)
            assignee.isAssingedToMe = true;
        if (assignee.checked) {
            game.addCharacter(assignee.name);
            if (assignee.isAssingedToMe)
                user.addCharacter(assignee.name);
        } else {
            game.removeCharacter(assignee.name);
            user.removeCharacter(assignee.name);
        }
        game.boardRenderer.redraw();
        if (playerChooser != null)
            playerChooser.postNotifyUpdateAssignee(assignee);
    }

    @Override
    public void onError(Exception e) {
        log.error(e);
        game.addPlayerComponentMessage("Error: " + e.getClass().getSimpleName() + ":" + e.getMessage());
    }

    @Override
    public UIZombicide getGameForUpdate() {
        return game;
    }

    @Override
    public void onGameUpdated(ZGame game) {
        this.game.boardRenderer.redraw();
        this.game.characterRenderer.redraw();
    }

    @Override
    public void onMessage(String msg) {
        game.addPlayerComponentMessage(msg);
    }

    @Override
    public void onDisconnected(String reason, boolean serverInitiated) {
        if (serverInitiated) {
            activity.game.setResult(null);
            if (playerChooser != null)
                playerChooser.dialog.dismiss();
            activity.runOnUiThread(new Runnable() {
                public void run() {
                    activity.newDialogBuilder().setTitle("Disconnected from Server")
                            .setMessage("Do you want to try and Reconnect?")
                            .setNegativeButton(R.string.popup_button_no, new DialogInterface.OnClickListener() {
                                @Override
                                public void onClick(DialogInterface dialog, int which) {
                                    activity.p2pShutdown();
                                }
                            }).setPositiveButton(R.string.popup_button_yes, new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {
                            client.reconnectAsync();
                        }
                    }).show();
                }
            });
        }
    }

    @Override
    public void onConnected() {
        client.register(ZUserMP.USER_ID, user);
        client.register(ZGameMP.GAME_ID, game);
        client.setDisplayName(activity.getDisplayName());
    }
}

File: zombicideandroid/src/main/java/cc/game/zombicide/android/ZClientMgr.kt
package cc.game.zombicide.android

import android.util.Log
import cc.game.zombicide.android.ZMPCommon.CL
import cc.lib.logger.LoggerFactory
import cc.lib.net.AGameClient
import cc.lib.net.GameCommand
import cc.lib.zombicide.ZGame
import cc.lib.zombicide.ZQuests
import cc.lib.zombicide.ZUser
import cc.lib.zombicide.p2p.ZGameMP
import cc.lib.zombicide.p2p.ZUserMP
import cc.lib.zombicide.ui.ConnectedUser
import cc.lib.zombicide.ui.UIZombicide
import java.util.Collections

/**
 * Created by Chris Caron on 7/28/21.
 */
class ZClientMgr(activity: ZombicideActivity, game: UIZombicide, override val client: AGameClient, val user: ZUser) :
	CL(activity, game), AGameClient.Listener, ZMPCommon.CLListener {

	var playerChooser: CharacterChooserDialogMP? = null
	val connectedPlayers = Collections.synchronizedList(mutableListOf<ConnectedUser>())

	init {
		addListener(this)
		client.setProperty("name", user.name)
		client.setProperty("color", user.colorId)
	}

	fun shutdown() {
		removeListener(this)
		client.removeListener(this)
	}

	override fun onCommand(cmd: GameCommand) {
		parseSVRCommand(cmd)
	}

	override fun onLoadQuest(quest: ZQuests) {
		game.loadQuest(quest)
		game.boardRenderer.redraw()
	}

	override fun onInit(color: Int, quest: ZQuests) {
		game.client = client
		game.setUserColorId(user, color)
		game.loadQuest(quest)
		activity.runOnUiThread {
			activity.initGameMenu()
		}
	}

	override fun onOpenAssignments(maxChars: Int, assignments: List<Assignee>) {
		showPlayerChooserDialog(maxChars, assignments)
	}

	override fun onMaxCharactersPerPlayerUpdated(max: Int) {
		activity.runOnUiThread {
			playerChooser?.updateMaxPlayers(max)
		}
	}

	fun showPlayerChooserDialog(maxCharacters: Int, playerAssignments: List<Assignee>) {
		activity.runOnUiThread {
			val assignees: MutableList<Assignee> = ArrayList()
			playerAssignments.forEach { assignee ->
				activity.charLocks.firstOrNull { assignee.name == it.player }?.let {
					assignee.lock = it
					assignees.add(assignee)
					if (assignee.isAssingedToMe)
						user.addCharacter(game.addCharacter(assignee.name))
				} ?: run {
					if (assignee.checked)
						game.addCharacter(assignee.name)
				}
			}
			playerChooser =
				object : CharacterChooserDialogMP(activity, assignees, false, maxCharacters) {
					override fun onAssigneeChecked(assignee: Assignee, checked: Boolean) {
						Log.d(TAG, "onAssigneeChecked: $assignee")
						val cmd = activity.clientMgr!!.newAssignCharacter(assignee.name, checked)
						object : CLSendCommandSpinnerTask(activity) {

							override fun onAssignPlayer(_assignee: Assignee) {
								if (assignee.name == _assignee.name)
									release()
							}

						override fun onSuccess() {
							assignee.checked = checked
							postNotifyUpdateAssignee(assignee)
						}
					}.execute(cmd)
				}

				override fun onStart() {
					dialog.dismiss()
					object : CLSendCommandSpinnerTask(activity) {

						override val timeout: Long = 30000L

						init {
							progressMessage = "Starting Game ...."
						}

						override fun onPlayerPressedStart(userName: String, numStarted: Int, numTotal: Int) {
							if (numStarted >= numTotal) {
								release()
								activity.runOnUiThread {
									activity.initGameMenu()
									activity.game.showQuestTitleOverlay()
								}
							} else
								publishProgress(arrayOf(numStarted, numTotal))
						}

						override fun onProgressUpdate(values: Array<Int>) {
							progressMessage = "${values[0]} of ${values[1]} have started"
						}

					}.execute(newStartPressed())
				}
			}
			game.boardRenderer.redraw()
		}
	}

	override fun onAssignPlayer(assignee: Assignee) {
		Log.d("ZClientMgr", "user.colorId: ${user.colorId} onAssignPlayer: $assignee")
		assignee.isAssingedToMe = user.colorId == assignee.color
		if (assignee.checked) {
			val c = game.addCharacter(assignee.name)
			if (assignee.isAssingedToMe) user.addCharacter(c)
		} else {
			game.removeCharacter(assignee.name)?.let {
				user.removeCharacter(it)
			}
		}
		game.boardRenderer.redraw()
		playerChooser?.postNotifyUpdateAssignee(assignee)
	}

	override fun onNetError(e: Exception) {
		log.error(e)
		game.addPlayerComponentMessage("Error: ${e.javaClass.simpleName} :  + ${e.message}")
	}

	override fun onGameUpdated(game: ZGame) {
		this.game.boardRenderer.redraw()
		this.game.characterRenderer.redraw()
	}

	override fun onMessage(msg: String) {
		game.addPlayerComponentMessage(msg)
	}

	override fun onDisconnected(reason: String, serverInitiated: Boolean) {
		activity.game.setResult(null)
		playerChooser?.dialog?.dismiss()
		activity.runOnUiThread {
			if (serverInitiated) {
				activity.p2pShutdown()
				activity.newDialogBuilder().setTitle("Disconnected from Server")
					.setMessage("Server Stopped: $reason")
					.setNegativeButton(R.string.popup_button_ok, null).show()
			} else activity.newDialogBuilder().setTitle("Disconnected from Server")
				.setMessage("Do you want to try and Reconnect?")
				.setNegativeButton(R.string.popup_button_no) { dialog, which -> activity.p2pShutdown() }
				.setPositiveButton(R.string.popup_button_yes) { dialog, which -> client.reconnectAsync() }
				.show()
		}
	}

	override fun onConnected() {
		client.register(ZUserMP.USER_ID, user)
		client.register(ZGameMP.GAME_ID, game)
		client.displayName = activity.displayName
	}

	override fun onConnectionsInfo(connections: List<ConnectedUser>) {
		connectedPlayers.clear()
		connectedPlayers.addAll(connections)
		activity.game.boardRenderer.redraw()
	}

	fun setColorId(id: Int) {
		client.setProperty("color", id)
	}

	companion object {
		var log = LoggerFactory.getLogger(ZClientMgr::class.java)
	}

	init {
		client.addListener(this)
	}
}